### DISCUSSION TITLE
Custom error bag ??

### DISCUSSION THREAD

**Discussion by mohamedsabil83** [livewire/livewire#2585] [1. Help]
Hi everyone.

I wonder if livewire accepts a custom error bag?

**e.g.** `Validator::make($attributes, $rules)->validateWithBag('custom_name')`

**Comment by lgmworks** [comment]
Currently not, but you can do the following

```php
try {
            $this->validate([
                'name' => 'required',
            ]);
        } catch (ValidationException $e) {
            session()->flash('userError', true);
            throw $e;
        }



@if(session('userError'))
    <div class="error-class">{{ $loopOverErrors }}</div>
@endif
```

**Reply by mohamedsabil83** [reply]
Thank you @luisgmoreno for the confirmation.
That's the workaround I currently used until @calebporzio supports the core bag customization.

**Reply by SimonMacIntyre** [reply]
Sorry to bump this old thing, it's the closest discussion and example to the problem I am having, possibly due to my own misunderstanding.

```php
public function updated($propertyName): void
    {
        try {
            $this->validateOnly($propertyName);
        } catch (ValidationException $e) {
            $errors = $this->getErrorBag(); //this is empty, to my surprise -- I thought it would have the failed validations already
                    if ($errors->has('some-field')) {
                        $errors->add('json', Arr::first($errors->get('some-field')));
                    }
                $this->setErrorBag($errors);
            throw $e;
        }
    }
```

Is there any way to customize the error bag in this way, from the validation failures? I essentially want to hook into quite complex nested array casts (from json), and spit their errors out into the singular high level `json` key in the errors. 

Does this mean there is something higher `upstream` in Livewire that catches this same error and does it in a later step? I'll try to go investigate the Livewire codebase to answer my own question in the meantime.

**Reply by SimonMacIntyre** [reply]
I was able to circumvent my very specific niche problem by just using stock Laravel Validator options:
```php
$validator->after(function (\Illuminate\Validation\Validator $validator) {
            foreach ($validator->errors()->get('json.*') as $message) {
                $validator->errors()->add(
                    'json', Arr::first($message)
                );
            }
        });
```

**Comment by joshhanley** [comment]
@mohamedsabil83 currently it won't work.

But can you explain why it would be beneficial and how you would make use of it?

**Reply by mohamedsabil83** [reply]
Hi @joshhanley, I guess it will give us more controling instead of changing the input name.

Let's say we are on the contact page which has an email input, also, there is an email input in the subscribe component on the same page. If the validation error appears for a contact email, it appears for one on the component.

I need to name the email input on the component different from one on the contact page to prevent that.

**Reply by joshhanley** [reply]
@mohamedsabil83 thanks for explaining 🙂

With how Livewire works, if you had both email fields within the same component, they would need different properties to bind too anyway, so there wouldn't be a collision there.

But if they were in two separate components, say contact component and subscribe component, both could have a property with the name of `email`, and both could use the default bag, and there still wouldn't be a collision.

The reason is that Livewire components run independently of each other, so each would use the default bag, but in separate network requests.

That is why I was unsure of the actual benefits of supporting custom error bags.

Hopefully the above makes sense. Let me know what you think and if I need to clarify anything 🙂

**Reply by mohamedsabil83** [reply]
@joshhanley what I mean is an email field in a livewire component and another one in the normal crud validation process e.g. send a contact form.

If contact validation returns an error message for the email field, it appears also for the email field in the livewire component.

It's not a big deal, I just faced that on one of my projects and just replace the normal contact form with a livewire component. Loved to share it with you and @calebporzio may it gives him hints for a future update.

Thank you all.

**Reply by joshhanley** [reply]
@mohamedsabil83 ah right, that makes sense! I didn't think about what would happen if you used a normal form with a Livewire form also on the page.

I guess as a work around could you do it the opposite way, and put the named bag on the normal form, and then use the default bag in the Livewire component? Not ideal, but could get it running.

I had a look at @devcircus suggestion below, I will reply to that with my thoughts.

**Comment by devcircus** [comment]
Is a custom error Bag what you're looking for? If so, I think components have an "errorBag" property. Just add 
```php
$this->errorBag = 'myCustomErrorBag';
```
to your component.

**Reply by joshhanley** [reply]
So I had a quick look into this, and Livewire actually stores the error bag object into that property, so assigning a string to that could potentially cause problems with it.

The code to set the error bag name is here, which is hardcoded, so I can't see a way around that currently.

https://github.com/livewire/livewire/blob/adaa7f9d822a781db01f921341b6d59c544c93da/src/Component.php#L176-L180

I can't imagine that'd be too hard to make that configurable though.

Hope this helps!

**Comment by mabdullahsari** [comment]
@joshhanley I hit this roadblock today as well. It'd be nice to allow this... I dug into the code and it seems like Livewire hardcodes the "default" error bag. 😕

**Reply by joshhanley** [reply]
@mabdullahsari can you have a look at [my comment above](https://github.com/livewire/livewire/discussions/2585#discussioncomment-1165540).

And then with that in mind, explain why it would be beneficial and how you would make use of it?

I personally don't have need for this, so would like to see a good use case for it 🙂

**Reply by mohamedsabil83** [reply]
I already [replied](https://github.com/livewire/livewire/discussions/2585#discussioncomment-1169432) to your mentioned comment.

A simple example: you have a contact page, which is a normal controller with request validation, and in the footer of the layout there is a livewire component for the subscription which has an **email** field. Let's say the visitor sent the contact form with wrongs in the **email** field that does not pass the validation rules, the validation message appears to both **email** fields in the contact form and subscription component.

I make a workaround for that and for other situations before. but in general, it's good if the error bag is not hardcoded in livewire.

**Reply by joshhanley** [reply]
Sorry @mohamedsabil83 I accidentally tagged the wrong person! I have edited my comment. I appreciated your answer and just hoping for another perspective from Muhammed 🙂

**Reply by mabdullahsari** [reply]
@joshhanley Yeah, sure thing. Context: e-commerce.

I have created a global `SlideOver` component which can render anything inside of it using a specific event name along with the name of the component that has to be rendered. So, something like this package: https://github.com/wire-elements/modal.

There is a product search page which displays around 100 products at any given time, and each has to be able to trigger their corresponding `Actions` (more on this in the paragraph below), so basically this was the right approach for me to take.

Now, whenever this `Actions` component has to come to the front inside the `SlideOver`, it has to display 3 collapsible sections:

- Add To Cart (default, expanded)
- Reserve Product
- Request Sample of Product

All of these have 2 fields in common: `quantity` and `batch`. **However**, they all trigger a **separate action method**. After discussing with my teammate, we decided to *not* persist these fields as class properties on the component, but instead manage the state in Alpine and pass them directly to their corresponding action methods within the `Actions` Livewire component which will then validate those parameters using a custom error bag. 

Simplified example:

```html
<div x-data="..."> <button x-on:click="$wire.call('addToCart', quantity, batch)">Add Product to Cart</button> </div>
<div x-data="..."> <button x-on:click="$wire.call('reserve', quantity, batch)">Reserve Product</button> </div>
<div x-data="..."> <button x-on:click="$wire.call('requestSample', quantity, batch)">Request Sample</button> </div>
```

As soon as an action has been run successfully, the `SlideOver` is closed automatically.

---

The way it is set up now works like a charm apart from the unfortunate error bag limitations. According to your example, to get the desired behaviour, I now have to create 5/6 separate classes:

- `Actions` component which will host the underlying, individual actions. The `SlideOver` can only render a single component at a time.
- `AddToCart` component
- `ReserveProduct` component
- `RequestSample` component
- A common `Validator` class / trait because they all have the same fields, but since they're split up into separate components now, I can't just declare a `protected` method and simply use that within the separate action methods.

IMHO, the cost of creating all of these extra components / traits is too much for this. Apart from the technicality, this is going to confuse our junior developers also as there will be 3 levels of nested components. The dev-tools unfortunately do not support nested components.

One could say that I can append all of the validated attributes with the action's name, something like:

```php
Validator::make(['quantityCart' => $quantity, 'batchCart' => $batch], $rules)->validate();
```
But now, I have to do extra work again to properly translate all of these because they deviate from the default translations found within the `attributes` key inside the `validation.php` translation file.

**Reply by mabdullahsari** [reply]
@joshhanley I'm definitely open to submitting a PR myself to allow this. However, I need to know all of the gotchas I'd need to keep in mind as there is probably a good reason for the hardcoded default error bag. If you can put me on the right track I'll hopefully make it till the end. Cheers.

**Reply by joshhanley** [reply]
@mabdullahsari thanks for the response! Yeah that all makes sense to me.

I don't think there should be any issues with the implementation, the examples are more just useful so when Caleb reviews it, he can have a look here to see why it's needed and how it would be used 🙂

Yeah feel free to submit a PR, and if you get stuck or not sure where to start, let me know and I can have a look 🙂

**Reply by mabdullahsari** [reply]
Alright, thanks a lot.

Here is an example of what I'd like to achieve post-fix: 

<img src="https://i.gyazo.com/19c3431893c24d9635b3373e3f33d0ff.gif" />

**Comment by Rattone** [comment]
Did this problem remain in v3 or is it possible to customize the bag error now? I do not think so

**Reply by joshhanley** [reply]
@Rattone I don't think anyone has PRd it, but as I mentioned above, PRs are welcome if someone who needs it wants to create it 🙂

**Reply by Rattone** [reply]
@joshhanley I don't know Livewire well enough to do that, anyway thanks for your time!

**Comment by katzda01** [comment]
I second what @joshhanley said, exactly the same problem - laravel's validation outer form's errors show up in every livewire component due to message bag not correctly implemented in livewire, cost me 1 day of a headage trying to find the non-existant proper solution due the the false impression that the $errorBag property (inside livewire validation) could be somehow set.

**Workaround is just hiding livewire's validation errors with a custom boolean flag showErrors = false; and only show when I interact directly with the livewire component.**

**Comment by Fluotonic** [comment]
There is a way but it might feel a bit _hacky_ for some. For me, not so much...

Here is what you need to do:

1. Create an intermediate model called **Component** or **CustomComponent** if you prefer
2. Inside this "custom component", copy and paste the output() method from Livewire/Component to this new component
3. Add a string $name public property to it
4. Edit the lines shown by @joshhanley to have **$this->name ?? 'default'** instead of just **'default'**.
5. Extend this new component from yours (instead of the Livewire one) and set your custom name in the mount() method
6. That's it! Now, you have named error bags by default :wink:

I know it's not perfect as it forces you to make sure your app doesn't break in future Livewire updates. But at least it allows us to get named error bags, which is very useful when you want to validate different forms (in multiple components) on the same page.

Anyway, I hope this will be helpful to anyone!
